File: Round1/src/company/snapchat/SpiralMatrix.java
package company.snapchat;

import java.util.ArrayList;
import java.util.List;

/**
 * 54. Given a matrix of m x n elements (m rows, n columns), return all elements of the matrix in spiral order.

For example,
Given the following matrix:

[
 [ 1, 2, 3 ],
 [ 4, 5, 6 ],
 [ 7, 8, 9 ]
]
 */
public class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<Integer>();
        if (matrix.length == 0) {
            return res;
        }
        
        int rowStart = 0;
        int rowEnd = matrix.length - 1;
        int colStart = 0;
        int colEnd = matrix[0].length - 1;
        
        while (rowStart <= rowEnd && colStart <= colEnd) {
            // Left to right on row
            for (int i = colStart; i <= colEnd; i++) {
                res.add(matrix[rowStart][i]);
            }
            rowStart++;
            
            // Top to bottom on col
            for (int i = rowStart; i <= rowEnd; i++) {
                res.add(matrix[i][colEnd]);
            }
            colEnd--;
            
            // Right to left on bottom row
            if (rowEnd >= rowStart) { // As rowStart is increasing, check if still need to
                for (int i = colEnd; i >= colStart; i--) {
                    res.add(matrix[rowEnd][i]);
                }
                rowEnd--;
            }
            
            // Bottom to top on left col
            if (colEnd >= colStart) {
                for (int i = rowEnd; i >= rowStart; i--) {
                    res.add(matrix[i][colStart]);
                }
                colStart++;
            }
        }
        
        return res;
    }
    
    /**
     * 59. Given an integer n, generate a square matrix filled with elements from 1 to n2 in spiral order.

		For example,
		Given n = 3,
		
		You should return the following matrix:
		[
		 [ 1, 2, 3 ],
		 [ 8, 9, 4 ],
		 [ 7, 6, 5 ]
		]
     */
    public int[][] generateMatrix(int n) {
        int[][] matrix = new int[n][n];
        
        int rowStart = 0;
        int rowEnd = n - 1;
        int colStart = 0;
        int colEnd = n - 1;
        
        int val = 1;
        while (rowStart <= rowEnd && colStart <= colEnd) {
            // Left to right on row
            for (int i = colStart; i <= colEnd; i++) {
                matrix[rowStart][i] = val++;
            }
            rowStart++;
            
            // Top to bottom on col
            for (int i = rowStart; i <= rowEnd; i++) {
                matrix[i][colEnd] = val++;
            }
            colEnd--;
            
            // Right to left on bottom row
            if (rowEnd >= rowStart) { // As rowStart is increasing, check if still need to
                for (int i = colEnd; i >= colStart; i--) {
                    matrix[rowEnd][i] = val++;
                }
                rowEnd--;
            }
            
            // Bottom to top on left col
            if (colEnd >= colStart) {
                for (int i = rowEnd; i >= rowStart; i--) {
                    matrix[i][colStart] = val++;
                }
                colStart++;
            }
        }
        
        return matrix;
    }
}

File: Round2/matrix/spiral_matrix.py
spiral_matrix.py


class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if not matrix:
            return matrix

        m, n = len(matrix), len(matrix[0])
        start_row = 0
        end_row = m - 1
        start_col = 0
        end_col = n - 1

        res = []
        while start_row <= end_row and start_col <= end_col:
            # go left to right on start row 
            for i in range(start_col, end_col + 1):
                res.append(matrix[start_row][i])
            start_row += 1

            # go top to bottom on end_col 
            for i in range(start_row, end_row + 1):
                res.append(matrix[i][end_col])
            end_col -= 1

            # go right to left on end_row 
            if end_row >= start_row:
                for i in range(end_col, start_col - 1, -1):
                    res.append(matrix[end_row][i])
                end_row -= 1

            # go bottom to top on start_col 
            if end_col >= start_col:
                for i in range(end_row, start_row - 1, -1):
                    print(i, start_col)
                    res.append(matrix[i][start_col])
                start_col += 1

        return res